src/biogeochem/ch4Mod.F90

```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
-                  k_h_cc = t_soisno(c,j) / k_h_inv * rgasLatm ! (4.21) Wania [(mol/m3w) / (mol/m3g)]
-                  oxdiffus = diffus_aere * d_con_g(2,1) / d_con_g(1,1) ! adjust for O2:CH4 molecular diffusion
-                  aerecond = area_tiller * rootfr(p,j) * oxdiffus / (z(c,j)*rob)
-                  aerecond = 1._r8 / (1._r8/(aerecond+smallnumber) + 1._r8/(grnd_ch4_cond(p)+smallnumber))
-                  oxaere = -aerecond *(conc_o2(c,j)/watsat(c,j)/k_h_cc - c_atm(g,2)) / dz(c,j) ![mol/m3-total/s]
-                  oxaere = max(oxaere, 0._r8)
-                  ! Diffusion in is positive; prevent backwards diffusion
-                  if ( .not. use_aereoxid_prog ) then ! fixed aere oxid proportion; will be done in ch4_tran
-                     oxaere = 0._r8
-                  end if
-               else
-                  aere = 0._r8
-                  oxaere = 0._r8
-               end if ! veg type, below water table, & above freezing
+            call SiteOxAere(is_vegetated, watsat(c,1:nlevsoi), h2osoi_vol(c,1:nlevsoi), t_soisno(c,1:nlevsoi), & 
+                 conc_ch4(c,1:nlevsoi), rootr(p,1:nlevsoi), qflx_tran_veg(p), jwt(c), &
+                 annsum_npp_ptr,annavg_agnpp_ptr, annavg_bgnpp_ptr, &
+                 elai(p), poros_tiller, rootfr_vr(1:nlevsoi), &
+                 grnd_ch4_cond(p), conc_o2(c,1:nlevsoi), c_atm(g,1:2), &
+                 z(c,1:nlevsoi), dz(c,1:nlevsoi), sat, & 
+                 tranloss(1:nlevsoi), &    ! Out
+                 aere(1:nlevsoi), &        ! Out
+                 oxaere(1:nlevsoi))        ! Out
 
+            do j = 1,nlevsoi
                ! Impose limitation based on available methane during timestep
                ! By imposing the limitation here, don't allow aerenchyma access to methane from other Patches.
-               aeretran = min(aere+tranloss, conc_ch4(c,j)/dtime + ch4_prod_depth(c,j))
+               aeretran = min(aere(j)+tranloss(j), conc_ch4(c,j)/dtime + ch4_prod_depth(c,j))
                ch4_aere_depth (c, j) = ch4_aere_depth(c,j) + aeretran*wtcol(p) ! patch weight in col.
-               ch4_tran_depth (c, j) = ch4_tran_depth(c,j) + min(tranloss, aeretran)*wtcol(p)
-               o2_aere_depth  (c, j) = o2_aere_depth (c,j) + oxaere*wtcol(p)
-            end do ! p filter
-         end do ! over levels
+               ch4_tran_depth (c, j) = ch4_tran_depth(c,j) + min(tranloss(j), aeretran)*wtcol(p)
+               o2_aere_depth  (c, j) = o2_aere_depth (c,j) + oxaere(j)*wtcol(p)
+            end do ! over levels
+
+         end do
       end if ! not lake
 
+      deallocate(tranloss)
+      deallocate(aere)
+      deallocate(oxaere)
+      deallocate(rootfr_vr)
+        
     end associate
 
   end subroutine ch4_aere
 
+  !--------------------------------------------------------------------------------------
+  
+  subroutine SiteOxAere(is_vegetated, & 
```

Comment:
We should keep in mind a little bit about efficiency. Compilers *should* optimize small single point subroutines like this. But, they don't always do a good job of it. When it's in the same module as in this case -- they should be able to though. But, we want to think about this anyway. Now, another reason to do this might be to for example add some unit testing around this SiteOxAere subroutine, as well as making it more modular. I actually think these later points are the most important, but we should always be mindful of efficiency.

Reply:
Agreed. On the fates side of the code, I'm trying to get our key subroutines that are most tied to the science (ie where the sauce is made, like leaf-level photosynthesis, etc) using primitive arguments and being calculated on a single element.  This makes the routine usable at different scales, presumably by different model scaling strategies (ie fates or big-leaf), and more readily unit testable.  However, I do see your point about efficiency. I wonder if there is a good fortran reference where the cover the performance trade-offs of encapsulating things in their own routines, vis-a-vis allocating/shuffling/copying/pointing to memory.